py/server/deephaven/dbc/__init__.py

```diff
@@ -2,34 +2,85 @@
-        conn (str): a connection string URI, please refer to https://sfu-db.github.io/connector-x/databases.html for
-            database specific format
+        conn (Any): must either be a connection string for the given driver or a Turbodbc/ADBC DBAPI Connection object
         query (str): SQL query statement
+        driver: (str): the driver to use, supported drivers are "odbc", "adbc", "connectorx", default is "connectorx"
 
     Returns:
         a new Table
 
     Raises:
         DHError
     """
-    try:
-        import connectorx as cx
-    except ImportError:
-        raise DHError(message="import ConnectorX failed, please install it first.")
+    if isinstance(conn, str):
+        if driver == "connectorx":
+            try:
+                import connectorx as cx
+            except ImportError:
+                raise DHError(message="import connectorx failed, please install it first.")
+
+            try:
+                pa_table = cx.read_sql(conn=conn, query=query, return_type="arrow")
+                return dharrow.to_table(pa_table)
+            except Exception as e:
+                raise DHError(e, message="failed to get a Arrow table from ConnectorX.") from e
+        elif driver == "odbc":
+            from deephaven.dbc.odbc import read_cursor
+            import turbodbc
+            with turbodbc.connect(connection_string=conn) as conn:
+                with conn.cursor() as cursor:
+                    cursor.execute(query)
+                    return read_cursor(cursor)
+        elif driver == "adbc":
+            from deephaven.dbc.adbc import read_cursor
+            if not conn:
+                import adbc_driver_sqlite.dbapi as dbapi
+            elif conn.strip().startswith("postgresql"):
+                import adbc_driver_postgresql.dbapi as dbapi
+            else:
+                raise DHError(message="not supported ADBC connection string")
+
+            with dbapi.connect(conn) as dbconn:
+                with dbconn.cursor() as cursor:
+                    cursor.execute(query)
+                    return read_cursor(cursor)
+        else:
+            raise DHError(message=f"unsupported driver {driver}")
+    else:
+        try:
+            import adbc_driver_manager.dbapi as dbapi
+            if isinstance(conn, dbapi.Connection):
+                from deephaven.dbc.adbc import read_cursor
+                with conn.cursor() as cursor:
+                    cursor.execute(query)
+                    return read_cursor(cursor)
+        except ImportError:
+            pass
+
+        try:
+            import turbodbc.connection
+            if isinstance(conn, turbodbc.connection.Connection):
+                from deephaven.dbc.odbc import read_cursor
+                with conn.cursor() as cursor:
+                    cursor.execute(query)
+                    return read_cursor(cursor)
+        except ImportError:
+            pass
+
+        raise DHError(message="invalid conn argument")
```

Comment:
```suggestion
        raise DHError(message="unsupported conn argument")
```
Should this print out the type of the conn arg?